### PR TITLE
Restore deleted internal registerJvmLanguageSourceDirectory API

### PR DESCRIPTION
**Pull request by tresat**
Fixes: #26547

Need to migrate it to JvmLanguageUtilities due to other project dependency changes.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/plugins/src/test/groovy/org/gradle/api/plugins/jvm/internal/DefaultJvmLanguageUtilitiesTest.groovy
Comment (jvandort) [inline subprojects/plugins/src/test/groovy/org/gradle/api/plugins/jvm/internal/DefaultJvmLanguageUtilitiesTest.groovy:1]: We probably also want to restore this integ tet:
https://github.com/gradle/gradle/blob/acf7091545e7a8a0c44bd93c7492c2f27f96f1ea/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/jpms/compile/JavaModuleCrossCompileIntegrationTest.groovy

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java [hunk 1]
```diff
@@ -36,4 +38,13 @@ public interface JvmLanguageUtilities {
      * @param compileTask the compile task which serves as reference for inference
      */
     <COMPILE extends AbstractCompile & HasCompileOptions> void useDefaultTargetPlatformInference(Configuration configuration, TaskProvider<COMPILE> compileTask);
+
+    /**
+     * Registers a new source directory for a source set, assuming that it will be compiled by
```
Comment (jvandort) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java:43]: Thoughts on adding some sort of disclaimer to this API, perhaps even an `@Deprecated`? 

We should be sure that users who _do_ use this are aware it is in fact internal and subject to change/removal/replacement

**Review by jvandort — APPROVED** [review]

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java (hunk 1 above)
Comment (tresat) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java:43]: I didn't do this in order to hold to the "don't deprecate without a replacement available" guidance.  I think we deprecate when we address the linked issue about making a public API.

Comment (tresat) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java:43]: It's in an `internal` package - that should be enough.  We've avoided deprecated other internal APIs in similar situations.

Comment (jvandort) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/JvmLanguageUtilities.java:43]: Ususally I'd say thats enough as well. Though the fact that this code is referenced in [Understanding Gradle](https://github.com/jjohannes/understanding-gradle) could confuse people. 

I'm fine leaving this as-is

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**